LoadBalancer for Kubernetes resulting in high number of requests to K8s API Server prior to every downstream call to lookup

**Describe the bug**
We have both Eureka discovery client (default) and spring-cloud-k8s  discovery client (**Fabric8** impl) in our spring apps that gets enabled based on environment. 

To make the switch from Eureka to K8s discovery we have following properties injected to all:
```
        - name: eureka.client.enabled
          value: "false"
        - name: spring.cloud.kubernetes.enabled
          value: "true"
        - name: spring.cloud.kubernetes.loadbalancer.mode
          value: "SERVICE"
```
When communicating from one service to another via LoadBalanced RestTemplate K8s API server is hit every time. This leads high load on K8s API resulting in rate limiting/blocking for the overall cluster.

 `spring.cloud.kubernetes.loadbalancer.mode` default to `POD` in which case its fine (K8s API server is not hit on each request ) but we wanted to "SERVICE" for our specific use case. 

A workaround is to disable loadbalancer in general when relying on K8s-Service using `spring.cloud.kubernetes.loadbalancer.enabled=false`, but this raises the question of usage of `loadbalancer.mode` other values.

Any insights on expected behaviour?

Versions - Happens both in spring-boot-3.0.2/spring-cloud-2022.0.1 and spring-boot-2.7.7/spring-cloud-2021.0.5

**Sample**
Sample code and process to reproduce locally has been pushed [here](https://github.com/dhananjay12/spring-microservice-demo/tree/master/k8s/spring-k8s-loadbalancer-test). The services involved in the sample K8s file source code is in the same repository.

I copied the class created the bean like
```
@Bean
	public ServiceInstanceListSupplier fabric8ServicesListSupplier(
			Environment environment, KubernetesClient kubernetesClient, Fabric8ServiceInstanceMapper mapper,
			KubernetesDiscoveryProperties discoveryProperties,
			ConfigurableApplicationContext context) {
		return ServiceInstanceListSupplier.builder().withBase(new Fabric8ServicesListSupplier(
						environment, kubernetesClient, mapper, discoveryProperties))
				.withCaching().build(context);
	}
```
On startup I get the following error:
```
2023-02-22T19:20:41.278Z ERROR 1 --- [           main[] o.s.boot.SpringApplication               : Application run failed
org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'fabric8ServicesListSupplier' defined in com.mynotes.microservices.demo.serviceone.ServiceOneApplication: Failed to instantiate [org.springframework.cloud.loadbalancer.core.ServiceInstanceListSupplier[]: Factory method 'fabric8ServicesListSupplier' threw exception with message: Name must be provided.
    at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:657) ~[spring-beans-6.0.4.jar:6.0.4[]
    at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:645) ~[spring-beans-6.0.4.jar:6.0.4[]
    at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1324) ~[spring-beans-6.0.4.jar:6.0.4[]
    at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1161) ~[spring-beans-6.0.4.jar:6.0.4[]
    at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:561) ~[spring-beans-6.0.4.jar:6.0.4[]
    at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:521) ~[spring-beans-6.0.4.jar:6.0.4[]
    at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:326) ~[spring-beans-6.0.4.jar:6.0.4[]
    at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234) ~[spring-beans-6.0.4.jar:6.0.4[]
    at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:324) ~[spring-beans-6.0.4.jar:6.0.4[]
    at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:200) ~[spring-beans-6.0.4.jar:6.0.4[]
    at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:961) ~[spring-beans-6.0.4.jar:6.0.4[]
    at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:915) ~[spring-context-6.0.4.jar:6.0.4[]
    at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:584) ~[spring-context-6.0.4.jar:6.0.4[]
    at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:146) ~[spring-boot-3.0.2.jar:3.0.2[]
    at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:730) ~[spring-boot-3.0.2.jar:3.0.2[]
    at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:432) ~[spring-boot-3.0.2.jar:3.0.2[]
    at org.springframework.boot.SpringApplication.run(SpringApplication.java:308) ~[spring-boot-3.0.2.jar:3.0.2[]
    at org.springframework.boot.SpringApplication.run(SpringApplication.java:1302) ~[spring-boot-3.0.2.jar:3.0.2[]
    at org.springframework.boot.SpringApplication.run(SpringApplication.java:1291) ~[spring-boot-3.0.2.jar:3.0.2[]
    at com.mynotes.microservices.demo.serviceone.ServiceOneApplication.main(ServiceOneApplication.java:34) ~[classes/:na]
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [org.springframework.cloud.loadbalancer.core.ServiceInstanceListSupplier[]: Factory method 'fabric8ServicesListSupplier' threw exception with message: Name must be provided.
    at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:171) ~[spring-beans-6.0.4.jar:6.0.4[]
    at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:653) ~[spring-beans-6.0.4.jar:6.0.4[]
    ... 19 common frames omitted
Caused by: java.lang.IllegalArgumentException: Name must be provided.
    at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.withName(BaseOperation.java:219) ~[kubernetes-client-6.2.0.jar:na[]
    at io.fabric8.kubernetes.client.dsl.internal.BaseOperation.withName(BaseOperation.java:88) ~[kubernetes-client-6.2.0.jar:na[]
```
Anything I am missing?

--- a/spring-cloud-kubernetes-client-loadbalancer/src/main/java/org/springframework/cloud/kubernetes/client/loadbalancer/KubernetesClientLoadBalancerClientConfiguration.java
+++ b/spring-cloud-kubernetes-client-loadbalancer/src/main/java/org/springframework/cloud/kubernetes/client/loadbalancer/KubernetesClientLoadBalancerClientConfiguration.java
@@ -21,7 +21,8 @@ import io.kubernetes.client.openapi.apis.CoreV1Api;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
 import org.springframework.cloud.kubernetes.commons.KubernetesNamespaceProvider;
 import org.springframework.cloud.kubernetes.commons.discovery.KubernetesDiscoveryProperties;
-import org.springframework.cloud.kubernetes.commons.loadbalancer.KubernetesServicesListSupplier;
+import org.springframework.cloud.loadbalancer.core.ServiceInstanceListSupplier;
+import org.springframework.context.ConfigurableApplicationContext;
 import org.springframework.context.annotation.Bean;
 import org.springframework.core.env.Environment;
 
@@ -32,11 +33,11 @@ public class KubernetesClientLoadBalancerClientConfiguration {
 
 	@Bean
 	@ConditionalOnProperty(name = "spring.cloud.kubernetes.loadbalancer.mode", havingValue = "SERVICE")
-	KubernetesServicesListSupplier kubernetesServicesListSupplier(Environment environment, CoreV1Api coreV1Api,
+	ServiceInstanceListSupplier kubernetesServicesListSupplier(Environment environment, CoreV1Api coreV1Api,
 			KubernetesClientServiceInstanceMapper mapper, KubernetesDiscoveryProperties discoveryProperties,
-			KubernetesNamespaceProvider kubernetesNamespaceProvider) {
-		return new KubernetesClientServicesListSupplier(environment, mapper, discoveryProperties, coreV1Api,
-				kubernetesNamespaceProvider);
+			KubernetesNamespaceProvider kubernetesNamespaceProvider, ConfigurableApplicationContext context) {
+		return ServiceInstanceListSupplier.builder().withBase(new KubernetesClientServicesListSupplier(environment, mapper, discoveryProperties, coreV1Api,
+				kubernetesNamespaceProvider)).withCaching().build(context);
 	}
 
 }

--- a/spring-cloud-kubernetes-fabric8-loadbalancer/src/main/java/org/springframework/cloud/kubernetes/fabric8/loadbalancer/Fabric8LoadBalancerClientConfiguration.java
+++ b/spring-cloud-kubernetes-fabric8-loadbalancer/src/main/java/org/springframework/cloud/kubernetes/fabric8/loadbalancer/Fabric8LoadBalancerClientConfiguration.java
@@ -20,7 +20,8 @@ import io.fabric8.kubernetes.client.KubernetesClient;
 
 import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
 import org.springframework.cloud.kubernetes.commons.discovery.KubernetesDiscoveryProperties;
-import org.springframework.cloud.kubernetes.commons.loadbalancer.KubernetesServicesListSupplier;
+import org.springframework.cloud.loadbalancer.core.ServiceInstanceListSupplier;
+import org.springframework.context.ConfigurableApplicationContext;
 import org.springframework.context.annotation.Bean;
 import org.springframework.core.env.Environment;
 
@@ -33,10 +34,10 @@ public class Fabric8LoadBalancerClientConfiguration {
 
 	@Bean
 	@ConditionalOnProperty(name = "spring.cloud.kubernetes.loadbalancer.mode", havingValue = "SERVICE")
-	KubernetesServicesListSupplier kubernetesServicesListSupplier(Environment environment,
+	ServiceInstanceListSupplier kubernetesServicesListSupplier(Environment environment,
 			KubernetesClient kubernetesClient, Fabric8ServiceInstanceMapper mapper,
-			KubernetesDiscoveryProperties discoveryProperties) {
-		return new Fabric8ServicesListSupplier(environment, kubernetesClient, mapper, discoveryProperties);
+			KubernetesDiscoveryProperties discoveryProperties, ConfigurableApplicationContext context) {
+		return ServiceInstanceListSupplier.builder().withBase(new Fabric8ServicesListSupplier(environment, kubernetesClient, mapper, discoveryProperties)).withCaching().build(context);
 	}
 
 }

